components/eppp_link/eppp_uart.c

```diff
@@ -0,0 +1,250 @@
+    uart_config.stop_bits = UART_STOP_BITS_1;
+    uart_config.flow_ctrl = UART_HW_FLOWCTRL_DISABLE;
+    uart_config.source_clk = UART_SCLK_DEFAULT;
+
+    ESP_RETURN_ON_ERROR(uart_driver_install(h->uart_port, config->rx_buffer_size, 0, config->queue_size, &h->uart_event_queue, 0), TAG, "Failed to install UART");
+    ESP_RETURN_ON_ERROR(uart_param_config(h->uart_port, &uart_config), TAG, "Failed to set params");
+    ESP_RETURN_ON_ERROR(uart_set_pin(h->uart_port, config->tx_io, config->rx_io, UART_PIN_NO_CHANGE, UART_PIN_NO_CHANGE), TAG, "Failed to set UART pins");
+    ESP_RETURN_ON_ERROR(uart_set_rx_timeout(h->uart_port, 1), TAG, "Failed to set UART Rx timeout");
+    return ESP_OK;
+}
+
+static void deinit_uart(struct eppp_uart *h)
+{
+    uart_driver_delete(h->uart_port);
+}
+
+#ifndef CONFIG_EPPP_LINK_USES_PPP
+/**
+ * @return size_t number of bytes left in the buffer
+ */
+static size_t process_packet(esp_netif_t *netif, uint8_t *buffer, size_t len)
+{
+    static uint8_t in_buf[2 * UART_BUF_SIZE] = {};
+    static size_t buf_start = 0;
+    static size_t buf_end = 0;
+    struct header *head;
+
+    // Add received data to our buffer
+    if (buf_end + len <= sizeof(in_buf)) {
+        memcpy(in_buf + buf_end, buffer, len);
+        buf_end += len;
+    } else {
+        ESP_LOGW(TAG, "Buffer overflow, discarding data");
+        buf_start = buf_end = 0;
+        return sizeof(in_buf) - buf_end;
+    }
+
+    // Process while we have enough data for at least a header
+    while ((buf_end - buf_start) >= sizeof(struct header)) {
+        head = (void *)(in_buf + buf_start);
+
+        if (head->magic != HEADER_MAGIC) {
+            goto recover;
+        }
+
+        uint8_t calculated_check = (head->size & 0xFF) ^ (head->size >> 8);
+        if (head->check != calculated_check) {
+            ESP_LOGW(TAG, "Checksum mismatch: expected 0x%04x, got 0x%04x", calculated_check, head->check);
+            goto recover;
+        }
+
+        // Check if we have the complete packet
+        uint16_t payload_size = head->size;
+        size_t total_packet_size = sizeof(struct header) + payload_size;
+
+        if (payload_size > MAX_PAYLOAD) {
+            ESP_LOGW(TAG, "Invalid payload size: %d", payload_size);
+            goto recover;
+        }
+
+        // If we don't have the complete packet yet, wait for more data
+        if ((buf_end - buf_start) < total_packet_size) {
+            ESP_LOGD(TAG, "Incomplete packet: got %d bytes, need %d bytes", (buf_end - buf_start), total_packet_size);
+            break;
+        }
+
+        // Got a complete packet, pass it to network
+        esp_netif_receive(netif, in_buf + buf_start + sizeof(struct header), payload_size, NULL);
+
+        // Advance start pointer past this packet
+        buf_start += total_packet_size;
+
+        // compact if we don't have enough space for 1x UART_BUF_SIZE
+        if (buf_start > (sizeof(in_buf) / 2) || (sizeof(in_buf) - buf_end) < UART_BUF_SIZE) {
```

Comment:
Just double check if the the watermark isn't set to low. If I understand it correctly it's half of the `in_buf`. Would it be possible to set it to e.g. 3/4?

Reply:
Reworked the receive function a bit, again, saving 1x`UART_BUF_SIZE`.

> So what about to #define UART_BUF_SIZE(ETH_MAX_PACKET_SIZE + AT_LEAST_HEADER_SIZE) to make it super clear.

Updated, just note that it's not `ETH_MAX_PACKET_SIZE`, but the maximum payload (which in this case is TUN packet size, not Ethernet)